### ISSUE TITLE
Memory consumption

### ISSUE THREAD

**Issue by andr0s** [jeanphix/Ghost.py#178]
I've played a bit with opening Yandex.ru and realized the memory consumption raises up significantly after you search something and click 'next page' a few times. Also, new threads are created (as I see from htop) - it was around 6 new threads, totally.

The standard's linux System Monitor shows that the Python process took 574 megabytes just to load a single webpage.

```
from ghost import Ghost
ghost = Ghost(wait_timeout=6660, display=True)
page, extra_resources = ghost.open("http://yandex.ru", wait=True)
ghost.capture_to('ghost.png')
result, resources = ghost.wait_for_alert()  # it's just for waiting and letting me control the browser with mouse and keyboard to perform searching
```

What's that and how can I reduce threads\consumed memory?

**Comment by cp2587** [comment]
@andreycizov Same user case here. Did you find a way to solve your problem ?

**Comment by andreycizov** [comment]
@cp2587 Never did - solved by restarting the celery daemon every hour. 

There are several problems with the idea: e.g. it uses a Qt thread singleton, as you may not have several of them running at the same time; the cleaning is only done by deleting the page object once (I guess, assuming that it will force the deletion of the whole chain of Qt objects), but I think there should be some sort of mechanism they're missing as well, which I did not get to.

**Comment by cp2587** [comment]
In celery you have a setting to restart your worker after a certain amount of executed tasks. May be useful to you.
Doing some test, i see the memory increase each time a open a new url (if i am staying on the same page doing javascript (even if there is ajax call) the memory is stationnary). I am not a qt user but seems to me the memory leak is in qt ? On the python side, i am calling del, .exit() and everything :(

**Comment by andreycizov** [comment]
Yes, thanks. I have anticipated that, too. It just didn't pop into my mind and I went with the restart of the celery process instead. Thanks.

Yeah, I have done that, too. In any case, I would firstly look into trying to kill every widget you have, or even trying to kill the thread singleton every time you do not have any tasks running in the browser at the current point in time. I would also look into Qt memory management - there might be a case they just try not to free some of the QObjects at all, because the runner process never quits, and a quick Google search leads [here](http://ralsina.me/weblog/posts/BB990.html), for example.

Do not think there's a memory leak in Qt per se, but I would assume there's something in PyQt that indeed does not agree with Python's approach to memory management.

**Comment by jeanphix** [comment]
@andreycizov @cp2587 This should be fixed in upcoming 0.2 https://github.com/jeanphix/Ghost.py/tree/0.2

Note that the API changed a bit, a new ``Session`` class has been introduced:

```python
import logging
from ghost import Ghost

ghost = Ghost(log_level=logging.DEBUG)

for i in xrange(0, 100):
    with ghost.start() as session:
        session.open('http://jeanphix.me')

ghost.exit()
```
or 
```python
for i in xrange(0, 100):
    session = ghost.start()
    session.open('http://jeanphix.me')
    session.exit()
```
